### DISCUSSION TITLE
Case action on next line

### DISCUSSION THREAD

**Discussion by waldo1001** [microsoft/alguidelines#90] [BC Best Practices]
Regarding: https://alguidelines.dev/bcbestpractices/readability/case-actions/

Personally, I'm not ok with this guideline.  I know the AL formatter automatically is going to format to:

```AL
    case Letter of
        'A':
            Letter2 := '10';
        'B':
            Letter2 := '11';
    end;
```

But for oneliners, I find this much more readable:

```AL
    case Letter of
        'A': Letter2 := '10';
        'B': Letter2 := '11';
    end;
```

What do you think?

**Comment by JeremyVyska** [comment]
For even the long one-liners, I find it WAY more readable with the 2nd format.

It also makes any multi-liner, begin/end, one stand out as a more complex scenario, and that adds instant value.

**Reply by GertHEctor** [reply]
Both are ok I think. First maybe better when a 'begin end' comes in and second maybe better when, as in the example, it involves 'one liners'

**Comment by mynavblog** [comment]
I am ok with any of the options but prefer option B. 

However my general rule is in such cases is and that I would write as a guideline: use AL formatter. 

I know it change from time to time and then the file is changed without any new code but I trust the formatter (I should not maybe but I can take that risk).

**Reply by waldo1001** [reply]
Thing is - I'd want the AL Formatter to react differently in some cases ;-)

**Reply by mynavblog** [reply]
I can agree with that - sometimes I also disagree with how it is formatting things. But then it is rather an idea/issue for the MS team and convinces them that something should change. Or I do not know if possible to have own Formatter as VS Code extension

**Comment by tinfister** [comment]
I prefer first option. But I wouldn't even mentioned if I see second option in the PR I'm reviewing, because is is personal preferences.
Both are clear and are doing the same job.
Do we need to put personal preferences in the best practices?

**Reply by NKarolak** [reply]
You've got a valid point here.

**Reply by waldo1001** [reply]
Well - it's a guideline for "readability" - so that's personal .. just trying to find out what's generally most readable.

But I agree - the world has bigger problems than this one .. ;-)

**Comment by pibcht** [comment]
I prefer 🅱️: OneLiners

Visual Studio gracefully solve this case by letting us define Code Style settings :
![image](https://user-images.githubusercontent.com/3841586/146184400-a6b4b7f7-9434-4e8d-a6e9-32fef86ad313.png)

Adding some VSCode extension settings could be a good improvement

**Reply by mortenseifert** [reply]
I am not fan of having the developer choose how to format the code - it should be the same for all developers on a projekt to produce clean code.

**Reply by TheDoubleH** [reply]
> I am not fan of having the developer choose how to format the code - it should be the same for all developers on a projekt to produce clean code.

100% - When we all write the code the same way, it makes it much easier to skim code.

**Reply by pibcht** [reply]
For me anything automatable (like formatting, tabs/spaces, LF/CRLF) should be left to the developer's discretion.
There should be a before-commit auto-format step with common standards, and also an after-pull auto-format with personal settings. 🤔
But I know it's a wishful thinking

**Comment by mortenseifert** [comment]
Actually I like both. First is more elegant and suits well when using begin...end - so if we have one option, I vote for the existing.

**Comment by vytjak** [comment]
Neither, but rather this:
```
case Letter of
    'A':
        begin
            Letter2 := '10';
        end;
    'B':
        begin
            Letter2 := '11';
        end;
end;
```
Yes, even for one liners. Then, if you need to add code to any of the cases, you will not need to add begin/end.

**Reply by waldo1001** [reply]
Interesting.  And what if another design-guideline could be: only use oneliners in casestatements - make sure to all out to individual functions ..?

**Reply by GertHEctor** [reply]
Could be a solution - and good readability and not many lines of code in procedures/functions

**Reply by TheDoubleH** [reply]
> Interesting. And what if another design-guideline could be: only use oneliners in casestatements - make sure to all out to individual functions ..?

Funny - I was actually thinking about that. And have on occasion done it, given the combined `CASE` statement easily ends up in a function that gets too long otherwise

**Comment by TheDenSter** [comment]
I've had too many issues with compound statements to teach myself to always use BEGIN/END even for one liners, and the only place I ever do is case statements. I would program it on single lines (and try to use function calls instead of programming the thing inside the case) and then the formatter would put them on separate lines. 

For case statements I don't have a clear preference, I think they are both very readable. Maybe the separate lines makes the case stand out a little more, so maybe that is preferable and why the formatter does it like that.

**Reply by TheDoubleH** [reply]
I like the Function Calls in the CASE.

**Reply by TheDenSter** [reply]
> I like the Function Calls in the CASE.

Yes me too, like in "No." OnValidate in the Sales Line table, I think that looks great, even with the function calls themselves on separate lines
![image](https://user-images.githubusercontent.com/23434319/146457582-cd870d4d-c990-4671-ba70-b689cd069548.png)

**Comment by frivadulla** [comment]
The first one, more readable IMO

**Comment by f4n0** [comment]
I prefer the onliner, but when saving the AL Formatter will change it to the first example..
